engine/src/main/java/org/terasology/engine/world/chunks/pipeline/ChunkProcessingPipeline.java

```diff
@@ -44,120 +32,162 @@
-                chunkProcessingInfo.endProcessing();
-                cleanup(chunkProcessingInfo);
+    private void processNewChunks(List<Chunk> chunks) {
+        for (Chunk chunk : chunks) {
+            Vector3ic position = chunk.getPosition();
+            if (chunkProcessingInfoMap.containsKey(position)) {
+                logger.error("DUPLICATE CHUNK!!!!");
             }
-            processChunkTasks();
-
-        } catch (ExecutionException e) {
-            String stageName =
-                    chunkProcessingInfo.getChunkTaskProvider() == null
-                            ? "Generation or Loading"
-                            : chunkProcessingInfo.getChunkTaskProvider().getName();
-            logger.error(
-                    String.format("ChunkTask at position %s and stage [%s] catch error: ",
-                            chunkProcessingInfo.getPosition(), stageName),
-                    e);
-            chunkProcessingInfo.getExternalFuture().setException(e);
-        } catch (CancellationException ignored) {
+
+            ChunkProcessingInfo chunkProcessingInfo = new ChunkProcessingInfo(position);
+            chunkProcessingInfo.setChunk(chunk);
+            chunkProcessingInfo.nextStage(stages);
+            chunkProcessingInfo.makeChunkTask();
+            chunkProcessingInfoMap.put(position, chunkProcessingInfo);
+        }
+
+        while (processChunkTasks()) {
+            continue;
         }
     }
 
-    private void processChunkTasks() {
-        chunkProcessingInfoMap.values().stream()
-                .filter(chunkProcessingInfo -> chunkProcessingInfo.getChunkTask() != null && chunkProcessingInfo.getCurrentFuture() == null)
-                .forEach(chunkProcessingInfo -> {
-                    org.terasology.engine.world.chunks.pipeline.stages.ChunkTask chunkTask = chunkProcessingInfo.getChunkTask();
-                    Set<Chunk> providedChunks = chunkTask.getRequirements().stream()
-                            .map(pos -> getChunkBy(chunkProcessingInfo.getChunkTaskProvider(), pos))
-                            .filter(Objects::nonNull)
-                            .collect(Collectors.toSet());
-                    if (providedChunks.size() == chunkTask.getRequirements().size()) {
-                        chunkProcessingInfo.setCurrentFuture(runTask(chunkTask, providedChunks));
+    /**
+     * @return whether any progress was made.
+     */
+    private boolean processChunkTasks() {
+        boolean anyChanged = false;
+        for (ChunkProcessingInfo chunkProcessingInfo : chunkProcessingInfoMap.values()) {
+            ChunkTask chunkTask = chunkProcessingInfo.getChunkTask();
+            if (chunkTask != null) {
+
+                List<Chunk> providedChunks = new ArrayList<>();
+                boolean satisfied = true;
+                for (Vector3ic pos : chunkTask.getRequirements()) {
+                    Chunk chunk = getChunkBy(chunkProcessingInfo.getChunkTaskProvider(), pos);
```

Comment:
what is the reasoning behind this. seems so strange to me. so we have 2 threads pushing chunks to the same concurrent map and at the same time we also attempt to iterate over the collection. there is going to be a lot of thread contention with all this as we lock and unlock this same map.